### DISCUSSION TITLE
Workbench Appearance Menu Rework

### DISCUSSION THREAD

**Discussion by MXKhronos** [HelixNebulaStudio/RiseOfTheDead#523] [Ideas]
Removal of default rust materials of low condition weapons.

With the new Appearance Menu rework, item conditions will unlock different tiers of customizations of your item.

---

### Customization Tiers

Battle-Tested:

- Color
- Skins

Heavily Worn:

- Part Transparency (0 to 1) (1=Hidden)

Ideal Model:

- Skin’s Texture Color, Offset, Scale and Transparency

Mint Condition:

- Part Offset

Factory New:

- Part Material (Will include most types of material such as Rust, Metal, Neon, Wood, Plastic, Slate, etc..)
- Part Reflectance (Only for part material set to plastic.)
- Skin Effects (Texture animations)

---

With more customization options, some QoL features are planned such as:

- More intuitive part selection by just clicking on the part directly to select.
- Customizing multiple parts at the same time.
- Copy and paste part configurations.

---

Since the new customizations will be saved differently compared to the old system, initial roll out of the feature is opt-in. When the new system is tested and stabilized, the old appearance menu will no long be accessible and your old customizations will be converted to the new system.

![Screenshot 2024-06-18 181309](https://github.com/HelixNebulaStudio/RiseOfTheDead/assets/38783502/cddabdb5-b6af-40c8-b3f9-ce52af880325)

![Screenshot 2024-06-18 181244](https://github.com/HelixNebulaStudio/RiseOfTheDead/assets/38783502/7d4e36ac-94b3-4804-b3f6-153744262a81)

**Comment by rafaderren** [comment]
With the rework of the appearance menu, will weapons below condition 0.15 able to be polished until its Factory New?

**Reply by MXKhronos** [reply]
Yes, all weapons/tools with capped conditions (e.g. `rusty48`) will be uncapped, so it can be polished to Factory New.

**Comment by MrGarrettFerret** [comment]
Looks very interesting, is there a way to toggle this in dev branch yet?

**Reply by MXKhronos** [reply]
Still in development at the moment, the screenshots are the UI draft, and only the backend data system is currently done. The label on the post will indicate `testing` when it's available for testing. I'll ping you if you want.

**Reply by MrGarrettFerret** [reply]
Alright, please do.

**Reply by MXKhronos** [reply]
New customization menu is now in dev branch, https://github.com/HelixNebulaStudio/RiseOfTheDead/commit/3d88c2e07d138efbab9588c41403fb7c60927349. @MrGarrettFerret

**Comment by Leroy8888** [comment]
This looks amazing honestly. The guns rust never really was the biggest issue in my eyes but more of a nuisance, but turning it from a generated number which influences (often hidden) aspects of guns customization to mutliple unlockable tiers seems very nice. Especially now that the ''lowest'' tier grants the basic additions.

I do have a question about the part offset, what exactly will this influence on a weapon or tool? Is this supposed to create some distance between parts? If so, what would the limit to it be?

**Reply by MXKhronos** [reply]
Yeah, you will be able to move parts around, e.g. if you want your weapon to look like it has a longer mag. It won't work well in every cases but that's up to you to design. Current limits are set to -1 to 1 units.

**Comment by Leroy8888** [comment]
I've thrown a quick glance at it and it looks very promising so far. There are still some issues I've encountered when I meddled with it, which I've left in the leaks channel of the unoff discord server (https://discord.com/channels/743948228571037787/754369005367853096/1257806992529948884).

I was also curious whether it would be possible to add a configuration to resize weapon parts. It wouldn't really need to increase the size to a massive scale (or at all), but it would be nice to rescale some parts to smaller size so that players can use existing models and morph them into their own ''weapon''.

For example, this revolver has mostly been modified with part offset, but the barrel does retrain its lenght and size. This kind of results in the parts clipping into each other and while it isn't as noticable on the example below, it could be on models like the rec-21.
![Schermopname (13078)](https://github.com/HelixNebulaStudio/RiseOfTheDead/assets/162458509/167f1764-14fc-48df-acca-52f204e904ca)
![Schermopname (13079)](https://github.com/HelixNebulaStudio/RiseOfTheDead/assets/162458509/793c4f21-016e-4ae7-a92f-c98e9968b74a)

Regarding the layers, is this feature finished yet? I can't really seem to use it properly unless I am doing it wrong.

**Comment by Leroy8888** [comment]
It also seems like the customization menu closes the moment you click anywhere but on the tool, so when you move a tool in the workbench to inspect it the menu often closes due to a misclick, which is quite the nuisance.

It would be nice if it remained open for the duration of the player customizing the tool.

**Comment by Leroy8888** [comment]
Also seems like the rusty texture / skins are exclusive to factory new range. Now the only issue with this is that prior to this change there actually are a handfull of people who want their weapons to be rusty for that kind of apocalyptic style, some might have even traded some gold for it.

So maybe the rusty materials should be available from the start, so that users with a battle tested weapons do not have to spend about 500 perks to polish the weapon or gamble their luck on building multiple instances of the same weapon.

Possibly the same for heavily worn if there are some materials that make it look worn down (but not completely corroded in acid)

**Comment by Leroy8888** [comment]
I've been thinking of suggesting part rotation, although at the time was a little bit unsure if would be a good idea. When proposed in the discord server, a valid point was brought up to where it could be used to create inappropiate imagery with weapon parts. Regardless of this, maybe it could still be possible to add this if it was for a *few* weapon parts, like the blades, barrels, scopes and or lenses of weapons. Perhaps it could also be included to more parts of the weapon, like the stock. 

It might be a bit overkill considering we already have quite some options when it comes to customization, but it would be yet another nice addition to it.

At the same time I would also like to suggest a small change to the light reflecting off weapons. Specifically present in live branch but also likely with the new costumization, light has a tendacy to reflect off weapon iron sights and such. This makes it rather difficult to use it in first person and it would be nice if the reflection was straight up disabled or decreased for these parts.
![Schermopname (13167)](https://github.com/HelixNebulaStudio/RiseOfTheDead/assets/162458509/03a2592b-f207-499a-a2c8-a657f981243b)

**Reply by MXKhronos** [reply]
I'll leave rotations adjustments for a future roadmap for now.

As for the flashlights, there's just no way to disable reflection. It's also a long time coming since I've wanted to adjust how weapon lighting works. The new light now comes from the weapon's bullet origin with a slight offset, so it should no longer be behind the sights now, but it'll basically move with the barrel if you're not aiming down sights.   Side note: it'll not originate from the weapon's "flashlight" just to maintain a good ADS experience.

**Comment by Leroy8888** [comment]
I've just noticed this now, but the help menu doesn't really mention how to create groups. It would be a nice small addition to include it in there somewhere.

The menu does state that multiple parts can be selected, but that's as far as it goes for groups. It also doesn't state that having multiple groups with the same name merges the parts when configuring either of the groups, despite it being obvious. However, this also has a tendacy to cause the modified menu to display the incorrect **amount** of parts included in the group.

Another request from a member would be to highlight the part when hovering over it, so that you can see what you're actually editing.

**Comment by JustDorien** [comment]
This new menu shouldn't be like "to customize weapon you need factory new 0.1" or "the weapon needs to be ideal model 0.3 to customize it" I find it pointless for players to polish weapons and wait 1h just to watch it improve by 1%, especially when they have good graph + battle tested
